File: easy/python/climbing_stairs.py
from typing import Dict

class Solution:
    def climbStairs(self, n: int, memo: Dict[int, int] = {}) -> int:
        if n <= 3:
            return n
        if n in memo:
            return memo[n]
        memo[n] = self.climbStairs(n - 1) + self.climbStairs(n - 2)
        return memo[n]

File: easy/ruby/climbing_stairs.rb
# @param {Integer} n
# @return {Integer}
def climb_stairs(n, memo = {})
  return n if n <= 3
  return memo[n] if memo[n]

  memo[n] = climb_stairs(n - 1, memo) + climb_stairs(n - 2, memo)
  memo[n]
end
